### PR TITLE
Implement project (Team) closing

### LINKED ISSUE #3602: Add project closing
Reticketed from #3601.

##### Notify

- [x] https://gratipay.freshdesk.com/helpdesk/tickets/2750
- [x] https://gratipay.freshdesk.com/helpdesk/tickets/2645
- [x] https://gratipay.freshdesk.com/helpdesk/tickets/6453
- [x] https://gratipay.freshdesk.com/helpdesk/tickets/6575
- [x] https://github.com/gratipay/gratipay.com/issues/4286

### PR DESCRIPTION
**Pull request by whit537**
See #3602.

- [x] set `is_closed`
- [x] ~~clear payment instructions (right?)~~ Wrong. Leave the data for now.
- [x] ~~clear takes (right?)~~ Wrong. Leave the data for now.
- [x] remove from payday
- [x] remove from homepage
- [x] remove from search
- [x] ~~are we showing 404?~~ No: 410.
- [x] ~~can you reopen a team?~~ No.
- [x] UI to close team
- [x] close teams automatically (with warning?) during participant closing?
- [x] ~~are we freeing up the name somehow?~~ No, not now.

### CODE REVIEW

**Comment by whit537** [comment]
Interesting. We [already have](https://github.com/gratipay/gratipay.com/blob/2035/sql/schema.sql#L416) an `is_closed` field. :-)

**Comment by whit537** [comment]
We [even](https://github.com/gratipay/gratipay.com/blob/2035/gratipay/billing/exchanges.py#L199-L211) [use](https://github.com/gratipay/gratipay.com/blob/2035/gratipay/billing/payday.py#L487) [it](https://github.com/gratipay/gratipay.com/blob/2035/gratipay/utils/__init__.py#L129-L130) (etc.)!

**Comment by whit537** [comment]
I don't understand how the edit form is wired.

**Comment by whit537** [comment]
We're using global vars and [`FormData`](http://caniuse.com/#feat=xhr2),

**Comment by whit537** [comment]
We're probably using FormData to [account](http://stackoverflow.com/questions/33469684/formdata-vs-serialize-what-is-the-difference) for the file upload.

**Comment by whit537** [comment]
Ready for review, @mattbk @nobodxbodon et al.

**Comment by mattbk** [comment]
Closing my own project as admin:
```
Traceback (most recent call last):
  File "/home/matt/Documents/gratipay/gratipay.com/env/lib/python2.7/site-packages/algorithm.py", line 288, in run
    new_state = function(**deps.as_kwargs)
  File "/home/matt/Documents/gratipay/gratipay.com/env/lib/python2.7/site-packages/aspen/algorithms/website.py", line 115, in get_response_for_resource
    return {'response': resource.respond(state)}
  File "/home/matt/Documents/gratipay/gratipay.com/env/lib/python2.7/site-packages/aspen/http/resource.py", line 59, in respond
    content_type, body = super(Dynamic, self).respond(accept, state)
  File "/home/matt/Documents/gratipay/gratipay.com/env/lib/python2.7/site-packages/aspen/simplates/__init__.py", line 169, in respond
    exec(self.pages[1], context)
  File "/home/matt/Documents/gratipay/gratipay.com/www/~/%username/index.html.spt", line 16, in <module>
    statement, stmt_lang = participant.get_statement(request.accept_langs)
  File "/home/matt/Documents/gratipay/gratipay.com/gratipay/models/participant/__init__.py", line 204, in get_statement
    """, dict(id=self.id, langs=langs), default=(None, None, None))
  File "/home/matt/Documents/gratipay/gratipay.com/env/lib/python2.7/site-packages/postgres/__init__.py", line 482, in one
    return cursor.one(sql, parameters, default)
  File "/home/matt/Documents/gratipay/gratipay.com/env/lib/python2.7/site-packages/postgres/cursors.py", line 105, in one
    out = self._some(sql, parameters, lo=0, hi=1)
  File "/home/matt/Documents/gratipay/gratipay.com/env/lib/python2.7/site-packages/postgres/cursors.py", line 127, in _some
    self.execute(sql, parameters)
  File "/home/matt/Documents/gratipay/gratipay.com/env/lib/python2.7/site-packages/psycopg2/extras.py", line 288, in execute
    return super(NamedTupleCursor, self).execute(query, vars)
ProgrammingError: column "content_scrubbed" does not exist
LINE 2:             SELECT content, content_scrubbed, lang
```

**Comment by mattbk** [comment]
Will make clean env and try again.

**Comment by mattbk** [comment]
But if I refresh, it says "already closed."

**Comment by whit537** [comment]
> column "content_scrubbed" does not exist

That suggests that you needed to `make schema` to pick up the new column from #4266.

**Comment by whit537** [comment]
> But if I refresh, it says "already closed."

Okay, so the close succeeded and the `content_scrubbed` error came afterwards?

**Comment by mattbk** [comment]
> Okay, so the close succeeded and the content_scrubbed error came afterwards?

Clicking "close" gave me the error, but the close succeeded. Getting back up to speed.

**Comment by mattbk** [comment]
Looks good from here!

File: www/%team/edit/close.spt [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+request.allow('POST')
+
+team = get_team(state)
+
+if not user.ADMIN and user.participant.username != team.owner:
+    raise Response(403, _("You are not authorized to access this page."))
+
+if team.is_closed:
+    raise Response(403, _("Already closed."))
```
Comment (nobodxbodon) [inline www/%team/edit/close.spt]: How about disable "close" button instead?

**Comment by mattbk** [comment]
Closed project is still listed on and accessible from ~user's profile.

**Comment by mattbk** [comment]
I hope to check through this again tonight and try to fix anything I find.

**Comment by mattbk** [comment]
> Closed project is still listed on and accessible from ~user's profile.
Fixxored in 6b191f0

I like it!

File: www/%team/edit/close.spt (hunk 1 above)
Comment (mattbk) [inline www/%team/edit/close.spt]: I think this was quicker to put together. We can enhance (add project reopening?) once the basic functionality is there.

Comment (nobodxbodon) [inline www/%team/edit/close.spt]: Sounds good. BTW after your latest commit (hide closed team), in theory user should not reach here, as they won't see the already closed teams?

Comment (mattbk) [inline www/%team/edit/close.spt]: If you try to close again by URL for some reason, you would get this error. All my latest commit did was hide the link to the project from the project owner's profile.

**Comment by whit537** [comment]
I added 36948c045fb2eebab59ea4ae3e5d89bcbe01e317 to improve (and test) our behavior around listing projects on owner profiles.

How does that look, @mattbk @nobodxbodon? Are we ready to merge?

**Comment by mattbk** [comment]
Good to me.

**Comment by whit537** [comment]
You good, @nobodxbodon?

**Comment by nobodxbodon** [comment]
@whit537 sorry I'm not sure I can review your latest commit  36948c0 today. Please go ahread and merge if LG to both of you?

BTW I feel the trickiness when a PR involves multiple peoples' commits. Ideally a third person who doesn't have commits should merge?

**Comment by mattbk** [comment]
If it makes you feel better, @whit537's commit overwrote mine :wink:.

**Comment by whit537** [comment]
It improved on yours. :-P

@mattbk The merge is yours, I think.

**Comment by whit537** [comment]
So is the delete branch? :eye:

**Comment by mattbk** [comment]
:stuck_out_tongue_closed_eyes:

**Pull request merged**